src/zkevm_specs/copy_circuit.py

```diff
@@ -0,0 +1,114 @@
+from typing import Dict, Iterator, List, NewType, Optional, Sequence, Union, Mapping, Tuple
+
+from .util import FQ, Expression, ConstraintSystem, cast_expr, MAX_N_BYTES, N_BYTES_MEMORY_ADDRESS
+from .evm import (
+    Tables,
+    CopyDataTypeTag,
+    CopyCircuitRow,
+    RW,
+    RWTableTag,
+    FixedTableTag,
+    CopyCircuit,
+    TxContextFieldTag,
+    BytecodeFieldTag,
+    TxLogFieldTag,
+)
+
+
+def lt(lhs: Expression, rhs: Expression, n_bytes: int) -> FQ:
+    assert n_bytes <= MAX_N_BYTES, "Too many bytes to composite an integer in field"
+    assert lhs.expr().n < 256**n_bytes, f"lhs {lhs} exceeds the range of {n_bytes} bytes"
+    assert rhs.expr().n < 256**n_bytes, f"rhs {rhs} exceeds the range of {n_bytes} bytes"
+    return FQ(lhs.expr().n < rhs.expr().n)
+
+
+def verify_row(cs: ConstraintSystem, rows: Sequence[CopyCircuitRow]):
+    cs.constrain_bool(rows[0].is_first)
+    cs.constrain_bool(rows[0].is_last)
+    # is_first == 0 when q_step == 0
+    cs.constrain_zero((1 - rows[0].q_step) * rows[0].is_first)
+    # is_last == 0 when q_step == 1
+    cs.constrain_zero(rows[0].q_step * rows[0].is_last)
+    cs.constrain_equal(rows[0].is_memory, cs.is_zero(rows[0].tag - CopyDataTypeTag.Memory))
+    cs.constrain_equal(rows[0].is_bytecode, cs.is_zero(rows[0].tag - CopyDataTypeTag.Bytecode))
+    cs.constrain_equal(rows[0].is_tx_calldata, cs.is_zero(rows[0].tag - CopyDataTypeTag.TxCalldata))
+    cs.constrain_equal(rows[0].is_tx_log, cs.is_zero(rows[0].tag - CopyDataTypeTag.TxLog))
+
+    # constrain the transition between two copy steps
+    is_last_two_rows = rows[0].is_last + rows[1].is_last
+    with cs.condition(1 - is_last_two_rows) as cs:
+        # not last two rows
+        cs.constrain_equal(rows[0].id, rows[2].id)
+        cs.constrain_equal(rows[0].log_id, rows[2].log_id)
+        cs.constrain_equal(rows[0].tag, rows[2].tag)
+        cs.constrain_equal(rows[0].addr + 1, rows[2].addr)
+        cs.constrain_equal(rows[0].src_addr_boundary, rows[2].src_addr_boundary)
+
+    # contrain the transition for `rw_counter` and `rwc_inc_left`
+    rw_diff = (1 - rows[0].is_pad) * (rows[0].is_memory + rows[0].is_tx_log)
+    with cs.condition(1 - rows[0].is_last) as cs:
+        # not last row
+        cs.constrain_equal(rows[0].rw_counter + rw_diff, rows[1].rw_counter)
+        cs.constrain_equal(rows[0].rwc_inc_left - rw_diff, rows[1].rwc_inc_left)
+    with cs.condition(rows[0].is_last) as cs:
+        # rwc_inc_left == 1 for last row in the copy slot
+        cs.constrain_equal(rows[0].rwc_inc_left, rw_diff)
+
+
+def verify_step(cs: ConstraintSystem, rows: Sequence[CopyCircuitRow]):
+    with cs.condition(rows[0].q_step):
+        # bytes_left == 1 for last step
+        cs.constrain_zero(rows[1].is_last * (1 - rows[0].bytes_left))
+        # bytes_left == bytes_left_next + 1 for non-last step
+        cs.constrain_zero((1 - rows[1].is_last) * (rows[0].bytes_left - rows[2].bytes_left - 1))
```

Comment:
not sure why use `rows[2].bytes_left` , refer to last step second row?

Reply:
If `rows[0]` is the read step (given that condition `rows[0].q_step`) then we want to constrain:
`rows[0].bytes_left == rows[2].bytes_left + 1` as `rows[2]` is the next read step if `rows[1]` was not the last step.